### ISSUE TITLE
protocol buffer compatibility

### ISSUE THREAD

**Issue by mildred** [ipfs/go-ipld#9]
There is something that puzzle me in `coding/pb`, it's the JSON format in which we deserialize the protocol buffer object. Whereas a unixfs would look like this in the new IPLD format:

```
{
  "file.txt": {
    mlink: "<hash>"
  }
}
```

The new format looks like this:

```
{
  links: [
    { name: "file.txt", hash: "<hash>", size: <size> }
  ]
  data: "<data">
}
```

The problem is that links will be completely different. Now, you can get the file at the path `/ipfs/<hash of pb object>/file.txt`, when we'll use IPLD, the file will no longer be accessible with the same path but rather the path `/ipfs/<hash of pb object>/files/0`.

How do you plan on making the JSON format you generate from the protocol buffer object compatible with the new IPLD format? I still don't get it.

**Comment by mildred** [comment]
There are few solutions for me:
- special case in the code the protobuf format (ugly and subject to collisions with native IPLD objects looking like protobuf generated objects)
- add handling in IPLD for this kind of structure (kind of a special handling, but generalized to other use cases as well, we really need to think what it could be)
- change the generated JSON from the protobuf object to something compatible with IPLD. I would imagine something like:
  
  ```
  {
    "file.txt": { mlink: "<hash>" },
    // Required to convert back to a protobuf object without losing the link order or any other information:
    @ipld-protobuf: {
      links: [
        { name: "file.txt", hash: "<hash>", size: <size> }
      ],
      data: "<data">
    }
  }
  ```

@jbenet What was your idea when you created the protobuf JSON format?

**Comment by jbenet** [comment]
yep you're right. i was thinking that we would either:
1. keep tweaking it to get it to be like what we just defined
2. make a special (go) type that overrides how `Links` are retrieved
3. or handle it specially in unixfs.
1. would be ideal, 2. and 3. are ok.

**Comment by mildred** [comment]
I'll try to make a proposition that goes with your first point here: We need to change the serialized format of protobuf objects so links don't change. I'll also try to make it easily compatible with JSON-LD.

What about serializing the object like this:

```
{
  @context: "/ipfs/<hash of context>"
  "file.txt": {
    order: 0, // We need to keep it to reconstruct the protobuf correctly
    mlink: "<hash>",
    size: <size>
  },
  @index: "files",
  @attrs: {
    data: "<data>",
    serialization: "/protobuf/msgio"
  }
}
```

I would imagine a context looking like this (it's not for processing within IPFS/IPLD but for other application to make use of the data, and show the format is compatible with JSON-LD):

```
{
  @type: ["IPFS/unixfs/Directory", "IPFS/unixfs/Entry"],
  files: {
    @container: @index,
    @index: "IPFS/unixfs/Entry/name", // JSON-LD extension
    @id: "IPFS/unixfs/Directory/hasFile",
    @type: "IPFS/unixfs/Entry",
    order: "IPFS/unixfs/Entry/order",
    size: "IPFS/unixfs/Entry/size",
    mlink: "IPFS/node/mlink"
  }
  data: "IPFS/unixfs/Entry/data",
  serialization: "IPFS/node/serializeTo"
}
```

(I'm not so sure about the way I wrote the context, it's to be verified)

**Comment by mildred** [comment]
In case we have links with duplicate names, we keep only one in the root object (the first or the last one, depends on how protobuf is implemented) and the extra links would appear elsewhere, for protobuf reconstruction only.

**Comment by jbenet** [comment]
@mildred that's pretty cool :+1: 

for order, would using an array like this work? (dont think it's better, just asking)

``` js
{
  ...
  @order: [
    "file.txt"
  ]
}
```

(though meh, tossup on which is more expensive.)

**Comment by jbenet** [comment]
(or some other non-clashing `@prefixed` thing

**Comment by mildred** [comment]
Why not, I initially thought it would be quite nice, but in the context of Linked Data/RDF object model, it doesn't makes sense. The reason is that the only collection in the Linked Data/RDF data model is the unordered set. So if we use a `@order` like directive and want it to be accessible to JSON-LD, we would have to duplicate it with the `order` key in each list item.

We can choose not to make it accessible to the JSON-LD model as the links were not meant to be ordered initially. The order is purely a side effect of protocol buffers items being ordered implicitly.

If we want to have a unique directive name, we could choose something like `@ipld-pb-order` instead.

**Comment by jbenet** [comment]
yeah

BTW, we can probably do ad-hoc `JSON --> JSON-LD` transformations if we make stackstream or just distribute code for those applications. I suppose users of `JSON-LD` might not object to running code with some guarantees (audited, signed, pure, resource bounded) to produce better representations of data. (not by default, i mean as an application on top of raw ipfs)

One simple example: supposed we have 4 objects: `J, J_ld, T, and A`:
- `J` is just some JSON data
- `J_ld` is the JSON-LD version of the data (produced by applying code in `T` to `J`)
- `T` contains a _transform_, code s.t. `T(J) --> J_ld`
- `A` "applies" `T(J) --> J_ld`, and links to `J`, `T`, (and maybe `J_ld`). (`J_ld` is virtual and can be either computed or stored, whatever the preference.)

**Comment by mildred** [comment]
Are you ok with a transformation from protobuf to JSON like this one?

```
{
  @context: "/ipfs/<hash of context>"
  "file.txt": {
    mlink: "<hash>",
    size: <size>
  },
  @index: "files",
  @attrs: {
    data: "<data>",
    serialization: "/protobuf/msgio"
  }
  @ipld-pb-order: [ "file.txt" ]
}
```

**Comment by jbenet** [comment]
@mildred this won't work. many protobuf links have no name. yes this is incompatible with our new assumptions and desires. yes this is annoying :/ 

we could do something like:

``` js
{
  "@links": [ <all the links here> ],
  "@data": "<the data here>",
  "foo": <link named "foo">,
  "bar": <link named "bar">,
}
```

meaning, have all links under `@links` but also patch the named links so resolving through works like new IPLD. yes this duplicates the JSON representation of the data. but we can probably be smart and dedup in the go and js implementations of it, as we'll be going  `protobuf <---> Go` and `protobuf <---> js`.

**Comment by mildred** [comment]
You're right, we should do something like that. I'd rather go for something like this instead:

``` js
{
  "@attrs": {
    "links": [ <all the links here> ],
    "data": "<the data here>",
  },
  "foo": <link named "foo">,
  "bar": <link named "bar">,
}
```

This way, we have compatibility with JSON-LD (the `links` and `data` keys can be described properly in a JSON-LD context)

**Comment by jbenet** [comment]
@mildred ok that works for me.

**Comment by mildred** [comment]
This has been implemented in PR #11. I'm closing this issue now. Reopen it if you want to further discuss it.
